### ISSUE TITLE
Create limit order with take profit and stop loss in Kucoin future with ccxt

### ISSUE THREAD

**Issue by MohammadTaherkhani** [ccxt/ccxt#12542]
Hi,
I want to create a limit order with take profit and stop loss in Kucoin future exchange.

For ex:
I created a long contract and I wanna close it when profit = %4 AND loss = %10 so, what are the params, types, etc.?

I tried with the below code but I think it's wrong.

Thank you.

- Windows:
- Programming Language version: Python
- CCXT version: 1.77.34

```
exchange = ccxt.kucoinfutures({
    'apiKey': config.MOH_BUY_API_KEY,
    'secret': config.MOH_BUY_API_SECRET,
    'password': config.MOH_BUY_API_PASSPHRASE
})

symbol = 'AVAX/USDT:USDT'
amount = 1
leverage = 10
side = 'buy'
price = 'Market close price for the symbol in 1 min candels ' #9.75
stopPrice = (100-config.STOPLOSS)*0.01*price

params = {'leverage': leverage, 'stop': 'loss', 'stopPrice': stopPrice , 'stopPriceType': 'TP'}

order = exchange.create_order(symbol=symbol, type='limit', side=side, amount=amount, price=price, params=params)

```

```
ccxt.base.errors.BadRequest: kucoinfutures error.createOrder.stopInvalid
```

**Comment by flav1ooo** [comment]
IDK why did u copy pasted my post lol https://github.com/ccxt/ccxt/issues/12533

**Comment by ttodua** [comment]
@MohammadTaherkhani please move your post as a comment into the referred issue (as @flav1ooo pointed out, this topic is duplicate).

**Comment by mefplusx** [comment]
> ttodua

this is differ exchanges. in binance and kucoinfutures hasn't similar options in params!
i have issue some like this. if kucoin support give me answer, im place description here

**Comment by YuxinSUN89** [comment]
Hi Folks, 

I am having the same issue here, but I didnt find the solution. 
Can anyone please advise ? I am not sure if its possible to add TP and SL in one go. 
Thanks

**Comment by mefplusx** [comment]
> Hi Folks,
> 
> I am having the same issue here, but I didnt find the solution :( Can anyone please advise ? Shout out to @ttodua and @carlosmiei
> 
> I am not sure if its possible to add TP and SL in one go. Thanks

this is possible in v2 API. but v2 now only in develop

**Comment by YuxinSUN89** [comment]
> 

Thanks @mefplusx  . 
OK that being said, we cannot create an order with TP and SL in one go (?)  We need to hard-coded to get them. 
Thanks

**Comment by ttodua** [comment]
I will take a look into this later and post an update.

**Comment by mefplusx** [comment]
@YuxinSUN89 @MohammadTaherkhani 
Guys! 
Kucoin give me answer!

Dear User,

Thank you for contacting us.

The Futures has two directions so they need to be discussed separately.

They all use the same API, but the input parameters are different.

 

POST /api/v1/orders

Param    type    Description

side    String    buy or sell

stop    String    [optional] Either down or up. Requires stopPrice and stopPriceType to be defined

stopPriceType    String    [optional] Either TP, IP or MP, Need to be defined if stop is specified.

stopPrice    String    [optional] Need to be defined if stop is specified.

 

1. when your side is buy, stop-loss and take-profit price set as follows

    1） stop-loss: side=buy, stop=down, stopPriceType and stopPrice are set according to your needs

    2) take-profit:side=buy, stop=up,  stopPriceType and stopPrice are set according to your needs

2. when your side is sell, stop-loss and take-profit price set as follows

    1） stop-loss: side=sell, stop=up, stopPriceType and stopPrice are set according to your needs

    2) take-profit: side=sell, stop=down, stopPriceType and stopPrice are set according to your needs

 

 

python demo:

       # Take Profit & Stop Loss

        order = self.client.create_limit_order(

            symbol='XBTUSDTM',

            side='buy',

            lever=1,

            clientOid=get_uuid(),

            size='1',

            price='55000',

            stop='down',

            stopPrice='56000',  # When the price is reached, it will enter the orderbook and the funds will be frozen

            stopPriceType='MP',

            remark='Its stop-profit is set at $ 55,000 (MP)',

        )

        order = self.client.create_limit_order(

            symbol='XBTUSDTM',

            side='buy',

            lever=1,

            clientOid=get_uuid(),

            size='1',

            price='61000',

            stop='up',

            stopPrice='60500',  # When the price is reached, it will enter the orderbook and the funds will be frozen

            stopPriceType='MP',

            remark='Its stop-loss is set at $ 61,000 (MP)',

        )

If you have any questions or concerns regarding this process, please don't hesitate to let us know.

Kind regards,

Aaron

KuCoin API Team

**Comment by mortezaataiy** [comment]
Hello, I think this code is not according to your opinion. You commented that if it is buy side and we want to set stop-loss then stop='down' but in the code you set it to 'up'. I think this code is wrong and below code is correct

```
# use stop as Take Profit

    order = self.client.create_limit_order(

        symbol='XBTUSDTM',

        side='buy',

        lever=1,

        clientOid=get_uuid(),

        size='1',

        price='55000',

        stop='up',

        stopPrice='56000',  # When the price is reached, it will enter the orderbook and the funds will be frozen

        stopPriceType='MP',

        remark='Its stop-profit is set at $ 56,000 (MP)',

    )

# use stop as Stop Loss
    order = self.client.create_limit_order(

        symbol='XBTUSDTM',

        side='buy',

        lever=1,

        clientOid=get_uuid(),

        size='1',

        price='61000',

        stop='down',

        stopPrice='60500',  # When the price is reached, it will enter the orderbook and the funds will be frozen

        stopPriceType='MP',

        remark='Its stop-loss is set at $ 60,500 (MP)',

    )
```
@mefplusx

**Comment by mortezaataiy** [comment]
CORRECT:

1）when your main order side is buy, stop-loss and take-profit order set as follows

    stop-loss: side=sell, stop=down, stopPriceType and stopPrice are set according to your needs

    take-profit:side=sell, stop=up, stopPriceType and stopPrice are set according to your needs

2）when your main order side is sell, stop-loss and take-profit order set as follows

    stop-loss: side=buy, stop=up, stopPriceType and stopPrice are set according to your needs

    take-profit: side=buy, stop=down, stopPriceType and stopPrice are set according to your needs
    
    
NOTIC: For stop-loss and take-profit order, reduceOnly: true

**Comment by chainoneer** [comment]
Thanks all of you for support.
But All of your answers are not right.
if you set stop parameter, then it creates stop orders, not TP & SL setting. 
if the stop price condition is satisfied, the order is filled and move to position, but no TP & SL setting is created.
I think there should be a new api.

**Comment by mortezaataiy** [comment]
Did you test it? @chainoneer 
Because I tested it by kucoin

**Comment by cserrb** [comment]
> Did you test it? @chainoneer Because I tested it by kucoin

I tested in poloniexfutures (same api that kucoinfutures) and it works perfectly. Thanks

**Comment by diego1979** [comment]
Good afternoon. Can I ask a question on this topic?

**Comment by 1J0A6L6** [comment]
In the latest CCXT documentation, it appears there are now take profit and stop loss orders across all exchanges. I'm waiting for my trading conditions to actually hit, but I'm curious whether the way the CCXT documentation has said it works in general is the way it works for Kucoin Futures (stoploss and takeprofit). Has anyone tested this with any good result? I don't want to think my code just isn't working.

**Comment by valbronx** [comment]
Hello!
Any news on this topic,
I can't find a way to nicely create an order with TP & SL,
Does someone manage it?
Thanks!!

**Comment by abestard** [comment]
Hello!

**Comment by ttodua** [comment]
I have researched Kucoin API docs and can say that it is not possible to create an order through API with attached SL & TP in one request.

You need to have an open position created at first, and only after that you can apply SL & TP orders in separate call. for that, you'll need to use https://docs.ccxt.com/#/README?id=conditional-orders 2nd type "StopLoss / TakeProfit" orders.

**Comment by ttodua** [comment]
closing the issue.
users are advised to check [`exchange.features`](https://docs.ccxt.com/#/README?id=features) to find out if any particular exchange supports embedded Stop-loss & Take-profit orders.
let us know if anyone faces any issues.
